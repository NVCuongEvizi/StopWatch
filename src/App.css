* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}

.App {
  text-align: center;
}

button {
  cursor: pointer;
  margin-right: 10px;
}

li {
  list-style-type: none;
  padding: 4px;
}

.time-button {
  width: 400px;
  margin: 40px auto 0;
  padding: 10px;
  background-color: aqua;

}

.time-button h2 {
  margin-bottom: 10px;
}

.time-button button {
  padding: 8px;
  border-radius: 3px;
  outline: none;
  border: none;
  font-size: 14px;
}

.time-button button:hover {
  background-color: violet;
}

.time-display {
  margin-top: 10px;
}