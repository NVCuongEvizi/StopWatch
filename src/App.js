import { useEffect, useRef, useState } from "react";
import "./App.css";

function App() {
  const [time, setTime] = useState(0);
  const [isClockRun, setIsClockRun] = useState(false);
  const [recordList, setRecordList] = useState([]);
  const intervalID = useRef(null);
  const timeStart = useRef(null);
  const timeRecord = useRef(0);

  useEffect(() => {
    return () => {
      clearInterval(intervalID.current);
    };
  }, []);

  const handleWatchRun = () => {
    if (!intervalID.current) {
      // start // resume
      intervalID.current = setInterval(() => {
        setTime((prevTime) => prevTime + 1);
      }, 1000);
      // console.log('start // resume')
    } else {
      // pause
      handleStop();
      // console.log('pause')
    }

    if (isClockRun) {
      // pause
      timeRecord.current += new Date().getTime() - timeStart.current;
      // console.log('pause')
    } else {
      // resume
      timeStart.current = new Date().getTime();
      // console.log('resume')
    }

    setIsClockRun(!isClockRun);
  };

  const handleReset = () => {
    timeStart.current = null;
    timeRecord.current = 0;

    setTime(0);
    setIsClockRun(false);
    setRecordList([]);
    handleStop();
  };

  const handleStop = () => {
    clearInterval(intervalID.current);
    intervalID.current = null;
  };

  const handleRecord = () => {
    const result =
      new Date().getTime() - timeStart.current + timeRecord.current;
    setRecordList((prevState) => [
      ...prevState,
      { watchTime: formatDate(), time: result / 1000 + "s" },
    ]);
  };

  const formatDate = () => {
    const today = new Date();
    let hour = today.getHours();
    let minute = today.getMinutes();
    let second = today.getSeconds();
    if (hour < 10) {
      hour = "0" + hour;
    }
    if (minute < 10) {
      minute = "0" + minute;
    }
    if (second < 10) {
      second = "0" + second;
    }
    return hour + ":" + minute + ":" + second;
  };

  return (
    <div className="App">
      <div className="time-button">
        <h2>{time}</h2>

        {timeRecord.current === 0 && !intervalID.current && (
          <button className="startBtn" onClick={handleWatchRun}>
            Start
          </button>
        )}

        <button className="resetBtn" onClick={handleReset}>
          Reset
        </button>

        <button className="pauseResumeBtn" onClick={handleWatchRun}>
          {isClockRun ? "Pause" : "Resume"}
        </button>

        {/* {isClockRun && (
          <button className="recordBtn" onClick={handleRecord}>
            Record
          </button>
        )} */}
        <button
          disabled={!isClockRun}
          className="recordBtn"
          onClick={handleRecord}
        >
          Record
        </button>
      </div>

      <ul className="time-display">
        <h2>Your time record:</h2>
        {recordList?.map((item, index) => (
          <li key={index}>
            #{index + 1} - watchTime: {item.watchTime} - Time: {item.time}
          </li>
        ))}
      </ul>
    </div>
  );
}

export default App;
