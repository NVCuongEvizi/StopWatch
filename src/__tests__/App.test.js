import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import App from "../App";

let wrapper;

beforeEach(() => {
  wrapper = shallow(<App />);
});

it("should match the snapshot", () => {
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

describe("render without error", () => {
  it("should has time-button and time-display", () => {
    expect(wrapper.find(".time-button")).toHaveLength(1);
    expect(wrapper.find(".time-display")).toHaveLength(1);
  });

  it("time start is zero", () => {
    expect(wrapper.find(".time-button h2").text()).toBe("0");
  });

  it("do not have list record when start", () => {
    expect(wrapper.find(".time-display").text()).toBe("Your time record:");
  });
});

describe("start button", () => {
  // fake time
  jest.useFakeTimers();

  it("time should up when click start button", () => {
    const startBtn = wrapper.find("button.startBtn");
    startBtn.simulate("click");

    // fake setTimeout
    jest.advanceTimersByTime(2000);

    const time = wrapper.find(".time-button h2").text();
    expect(time).toBe("2");
  });
});

describe("reset button", () => {
  jest.useFakeTimers();

  it("time should return zero when click reset", () => {
    const startBtn = wrapper.find("button.startBtn");
    const resetBtn = wrapper.find("button.resetBtn");
    startBtn.simulate("click");

    jest.advanceTimersByTime(4000);

    resetBtn.simulate("click");
    const time = wrapper.find(".time-button h2").text();
    expect(time).toBe("0");
  });
});

describe("pause button", () => {
  jest.useFakeTimers();

  it("time pause when click pause button", () => {
    const startBtn = wrapper.find("button.startBtn");
    const pauseBtn = wrapper.find("button.pauseResumeBtn");
    const recordBtn = wrapper.find("button.recordBtn");
    startBtn.simulate("click");

    jest.advanceTimersByTime(2000);

    pauseBtn.simulate("click");

    jest.advanceTimersByTime(3000);
    recordBtn.simulate("click");
    const time = wrapper.find(".time-button h2").text();
    expect(time).toBe("2");
  });
});

describe("record button", () => {
  jest.useFakeTimers();

  it("get recordList when click record button", () => {
    const startBtn = wrapper.find("button.startBtn");
    const recordBtn = wrapper.find("button.recordBtn");
    startBtn.simulate("click");

    jest.advanceTimersByTime(3000);

    const recordBtnClick_times = 5;
    for (let i = 0; i < recordBtnClick_times; i++) {
      recordBtn.simulate("click");
    }
    const recordList = wrapper.find(".time-display li");
    expect(recordList).toHaveLength(recordBtnClick_times);
  });

  it("get right value when first click record button", () => {
    const startBtn = wrapper.find("button.startBtn");
    const recordBtn = wrapper.find("button.recordBtn");
    startBtn.simulate("click");

    jest.advanceTimersByTime(7000);

    recordBtn.simulate("click");
    const time = wrapper.find(".time-button h2").text();
    expect(time).toBe("7");
    const recordList = wrapper.find(".time-display li");
    expect(recordList).toHaveLength(1);
  });

  it("get right value when click pause and reset one times", () => {
    const startBtn = wrapper.find("button.startBtn");
    const pauseResumeBtn = wrapper.find("button.pauseResumeBtn");
    const recordBtn = wrapper.find("button.recordBtn");
    startBtn.simulate("click");

    jest.advanceTimersByTime(2000);

    pauseResumeBtn.simulate("click");
    jest.advanceTimersByTime(4000);
    pauseResumeBtn.simulate("click");
    jest.advanceTimersByTime(1000);

    recordBtn.simulate("click");
    const time = wrapper.find(".time-button h2").text();
    expect(time).toBe("3");
    const recordList = wrapper.find(".time-display li");
    expect(recordList).toHaveLength(1);
  });

  it("get right value when click pause and reset two times", () => {
    const startBtn = wrapper.find("button.startBtn");
    const pauseResumeBtn = wrapper.find("button.pauseResumeBtn");
    const recordBtn = wrapper.find("button.recordBtn");
    startBtn.simulate("click");

    jest.advanceTimersByTime(2000);

    pauseResumeBtn.simulate("click");
    jest.advanceTimersByTime(4000);
    pauseResumeBtn.simulate("click");
    jest.advanceTimersByTime(1000);

    pauseResumeBtn.simulate("click");
    jest.advanceTimersByTime(3000);
    pauseResumeBtn.simulate("click");
    jest.advanceTimersByTime(2000);

    recordBtn.simulate("click");
    const time = wrapper.find(".time-button h2").text();
    expect(time).toBe("5");
    const recordList = wrapper.find(".time-display li");
    expect(recordList).toHaveLength(1);
  });
});
