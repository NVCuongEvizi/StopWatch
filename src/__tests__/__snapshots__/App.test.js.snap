// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="App"
>
  <div
    className="time-button"
  >
    <h2>
      0
    </h2>
    <button
      className="startBtn"
      onClick={[Function]}
    >
      Start
    </button>
    <button
      className="resetBtn"
      onClick={[Function]}
    >
      Reset
    </button>
    <button
      className="pauseResumeBtn"
      onClick={[Function]}
    >
      Resume
    </button>
    <button
      className="recordBtn"
      disabled={true}
      onClick={[Function]}
    >
      Record
    </button>
  </div>
  <ul
    className="time-display"
  >
    <h2>
      Your time record:
    </h2>
  </ul>
</div>
`;
